libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/ReferenceCollector.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+#include <libsolutil/Common.h>
+
+#include <typeinfo>
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+ReferenceCollector::ReferenceCollector(
+	Declaration const& _declaration,
+	string const& _sourceIdentifierName
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _astSearchRoot,
+	string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_astSearchRoot.accept(collector);
+	return std::move(collector.m_collectedReferences);
+}
+
+vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	int _sourceOffset,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto references = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+		references += collect(identifier->annotation().referencedDeclaration, _sourceUnit, identifier->name());
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		for (size_t i = 0; i < identifierPath->pathLocations().size(); ++i)
+		{
+			SourceLocation const location = identifierPath->pathLocations()[i];
+			if (location.containsOffset(_sourceOffset))
+			{
+				Declaration const* declaration = identifierPath->annotation().pathDeclarations.at(i);
+				ASTString const& name = identifierPath->path().at(i);
+				references += collect(declaration, _sourceUnit, name);
```

Comment:
All these `+=` look like you want to keep adding stuff to `references` but you only ever add one thing and then immediately return. Wouldn't it be simpler to just `return collect(...)` in all these `if`s?

Reply:
What about the other `return`s in this function? I mean, with this one place changed it's not as much of a simplification as I was hoping for :) I think you can just remove `references` completely.

Reply:
@cameel applied and squashed.